web/src/message_view.js

```diff
@@ -678,6 +719,18 @@ export function show(raw_terms, opts) {
         assert(select_opts !== undefined);
         assert(select_immediately !== undefined);
 
+        // We don't rely on the existing `id_info.final_select_id`
+        // for `with` operator because, `with` operator requires
+        // last unread. Due to "retarget message location" in
+        // case of old permalinks, the show() method would be
+        // called an extra time for change_hash, and if user has
+        // `mark_read_on_scroll_policy` set, he will end up at last
+        // message rather.
+        if (filter.has_operator("with") && id_info.final_select_id === LARGER_THAN_MAX_MESSAGE_ID) {
+            id_info.final_select_id = last_unread_msg_id;
+            last_unread_msg_id = LARGER_THAN_MAX_MESSAGE_ID;
+        }
+
```

Comment:
This section is important because sometimes when `message_view.show()` gets called more than once such as when `retarget message location` condition, users with `mark_read_on_scroll_policy` will end up at last message rather than last unread message.

Any thoughts on a better way?

Reply:
@amanagr could you have a look at this when you're free.

Reply:
Removed this code, so no longer relevant.